specs/state-proof.md

```diff
@@ -0,0 +1,351 @@
+# State Proof
+
+In EVM the interpreter has ability to do any random read-write access to data like account balance, account storage, or stack and memory in current scope, but it's hard for a circuit to keep tracking these data to ensure their consistency from time to time. So the state proof helps EVM proof to check all the random read-write access records are valid, throgh grouping them by their unique index first, and then sorting them by accessing timestamp. We call the accessing timestamp **global counter**, which counts the number of access records and also serves as an unique identifier for a record.
+
+When state proof is generated, a random access record set is also produced with the global counter as each record's unique identifier, we called this set **bus mapping** because it acts like the bus in computer which transfers data between components. The bus mapping will be shared to EVM proof, then EVM proof can do random read-write access to anything inside the set at any time, with extra subset arguments to argue those accesses indeed form a subset of the one prodced by state proof.
+
+## Random Read-Write Data
+
+State proof maintains the read-write part of [random accessible data](./evm-proof.md#Random-Accessible-Data) of evm proof:
+
+| Target                                | Index             | Description                              |
+| ------------------------------------- | ----------------- | ---------------------------------------- |
+| [`AccountNonce`](#AccountNonce)       | `{address}`       | Account's nonce                          |
+| [`AccountBalance`](#AccountBalance)   | `{address}`       | Account's balance                        |
+| [`AccountCodeHash`](#AccountCodeHash) | `{address}`       | Account's code hash                      |
+| [`AccountStorage`](#AccountStorage)   | `{address}.{key}` | Account's storage as a key-value mapping |
+| [`CallState`](#CallState)             | `{id}.{enum}`     | Call's internal state                    |
+| [`CallStateStack`](#CallStateStack)   | `{id}.{index}`    | Call's stack as a encoded word array     |
+| [`CallStateMemory`](#CallStateMemory) | `{id}.{index}`    | Call's memory as a byte array            |
+
+The concatenation of **Target** and **Index** becomes the unique index for data. Each record will be attached with a global counter as their accessing timestamp, and the records are constraint to be in group by their unique index first and to be sorted by their global counter increasingly. Given the access to previous record, each target has their custom constraints, for example, `AccountNonce` always increase by 1, and values in `CallStateMemory` should fit in 8-bit.
```

Comment:
> each target has their custom constraints

Is the "custom constraints" here refers to the one we use in the circuit? Or do we want to say each target has its specific format and rules to update?

Reply:
> The concatenation of **Target** and **Index** becomes the unique index for data

We might want to define the concrete concatenation inputs and outputs later

Reply:
> each target has their custom constraints

We want to say each target has its specific format and rules to update. Maybe we should say:

> each target has their own format and rules to update.

Reply:
> We might want to define the concrete concatenation inputs and outputs later

For sure, I left it ambiguous because it could be encoded with tricky optimization. Let me add a reminder here.